connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/DeadLetterQueueReporter.java

```diff
@@ -118,11 +140,73 @@ public void report(ProcessingContext context) {
                     originalMessage.key(), originalMessage.value(), originalMessage.headers());
         }
 
+        if (connConfig.isDlqContextHeadersEnabled()) {
+            populateContextHeaders(producerRecord, context);
+        }
+
         this.kafkaProducer.send(producerRecord, (metadata, exception) -> {
             if (exception != null) {
                 log.error("Could not produce message to dead letter queue. topic=" + dlqTopicName, exception);
                 errorHandlingMetrics.recordDeadLetterQueueProduceFailed();
             }
         });
     }
+
+    // Visible for testing
+    void populateContextHeaders(ProducerRecord<byte[], byte[]> producerRecord, ProcessingContext context) {
+        String topic = "";
+        int partition = -1;
+        long offset = -1;
+        if (context.consumerRecord() != null) {
+            topic = context.consumerRecord().topic();
+            partition = context.consumerRecord().partition();
+            offset = context.consumerRecord().offset();
+        }
+
+        add(producerRecord.headers(), prefix(ERROR_HEADER_ORIG_TOPIC), topic);
+        add(producerRecord.headers(), prefix(ERROR_HEADER_ORIG_PARTITION), String.valueOf(partition));
+        add(producerRecord.headers(), prefix(ERROR_HEADER_ORIG_OFFSET), String.valueOf(offset));
+        add(producerRecord.headers(), prefix(ERROR_HEADER_CONNECTOR_NAME), connectorTaskId.connector());
+        add(producerRecord.headers(), prefix(ERROR_HEADER_TASK_ID), String.valueOf(connectorTaskId.task()));
+        add(producerRecord.headers(), prefix(ERROR_HEADER_STAGE), context.stage().name());
+        add(producerRecord.headers(), prefix(ERROR_HEADER_EXECUTING_CLASS), context.executingClass().getName());
```

Comment:
How about relying more upon the `ConnectHeaders` methods:
```
Headers headers = producerRecord.headers();
headers.add(ERROR_HEADER_ORIG_TOPIC, asBytes(topicName));
headers.add(ERROR_HEADER_ORIG_PARTITION, asBytes(partition));
...
```
and then simply overloading the `asBytes(...)` method for the various types of values, including `int`, `String`, etc., and even `Throwable`?